### ISSUE TITLE
generalized dice loss

### ISSUE THREAD

**Issue by DecentMakeover** [wolny/pytorch-3dunet#29]
Hi 

I have borrowed your implementation of generalized dice loss for my custom network(Thanks for sharing your work), but i am training a multi label segmentation task , will it work, or do i have to modify it ?

Thanks

**Comment by wolny** [comment]
Hi @DecentMakeover,

happy to hear that you find some of my implementation useful.
Regarding your question: yes, it will work but make sure to specify **sigmoid_normalization=False** when constructing the loss instance (see: https://github.com/wolny/pytorch-3dunet/blob/master/unet3d/losses.py#L79) since you have a multi-label segmentation task and thus you need Softmax normalization. Using Dice loss for non-binary segmentation tasks (i.e. multi-label like in your case) is non-standard since the Dice coefficient is a measure of overlap between two sets, but I used it for multi-label segmentation in a way described above with satisfactory results. Make sure to compare it with the standard cross-entropy loss though which is a default choice for multi-label segmentation problems.

**Comment by DecentMakeover** [comment]
but i thought dice loss was standard in segmentation tasks, do you think standard cross entropy would be a better choice for multi label segmentation problem? or do you have a different loss function you would like to suggest?

**Comment by wolny** [comment]
> but i thought dice loss was standard in segmentation tasks

only when it's about foreground/background

> do you think standard cross entropy would be a better choice for multi label segmentation problem?

yes, but try both

> Also could you kindly guide me as to how to use the weight argument in the loss function?

`weight` has the same semantic as for the CrossEntropy loss, see: https://pytorch.org/docs/stable/nn.html#crossentropyloss

**Comment by DecentMakeover** [comment]
Thanks a lot

**Comment by DecentMakeover** [comment]
@wolny 

Hey, my loss is near 0 but predictions are very messy, for every pixel the model predicts some class label from 0-24, i mean most pixels have value from 1-24 instead of being mostly 0s.

the question i had is, i am converting the labels to one hot vector for the dice loss, that shouldnt cause any issue right?

**Comment by wolny** [comment]
> the question i had is, i am converting the labels to one hot vector for the dice loss, that shouldnt cause any issue right?

no it shouldn't, cause the loss function does expect the labels to be as one hot vector and then it computes the Dice on a per channel basis

**Comment by DecentMakeover** [comment]
okay thanks

**Comment by DecentMakeover** [comment]
@wolny 

Hey do you want to team up ? 

https://verse2019.grand-challenge.org/

please let me know
